source/postprocess/global_statistics.cc

```diff
@@ -0,0 +1,296 @@
+      // with the first time step
+      if (!this->get_parameters().run_postprocessors_on_initial_refinement)
+        this->get_signals().post_set_initial_state.connect(std_cxx11::bind(&aspect::Postprocess::GlobalStatistics<dim>::clear_data,
+                                                                           std_cxx11::ref(*this)));
+    }
+
+
+    template <int dim>
+    void
+    GlobalStatistics<dim>::clear_data()
+    {
+      list_of_S_iterations.clear();
+      list_of_A_iterations.clear();
+      solver_controls_cheap.clear();
+      solver_controls_expensive.clear();
+      advection_solver_controls.clear();
+    }
+
+
+    template <int dim>
+    void
+    GlobalStatistics<dim>::store_stokes_solver_history(const unsigned int number_S_iterations,
+                                                       const unsigned int number_A_iterations,
+                                                       const SolverControl &solver_control_cheap,
+                                                       const SolverControl &solver_control_expensive)
+    {
+      list_of_S_iterations.push_back(number_S_iterations);
+      list_of_A_iterations.push_back(number_A_iterations);
+      solver_controls_cheap.push_back(solver_control_cheap);
+      solver_controls_expensive.push_back(solver_control_expensive);
+    }
+
+    template <int dim>
+    void
+    GlobalStatistics<dim>::store_advection_solver_history(const bool solved_temperature_field,
+                                                          const unsigned int compositional_index,
+                                                          const SolverControl &solver_control)
+    {
+      const std::string column_name = (solved_temperature_field) ?
+                                      "Iterations for temperature solver"
+                                      :
+                                      "Iterations for composition solver " + Utilities::int_to_string(compositional_index+1);
+
+      unsigned int column_position = numbers::invalid_unsigned_int;
+
+      for (unsigned int i=0; i<advection_solver_controls.size(); ++i)
+        if (column_name == advection_solver_controls[i].first)
+          column_position = i;
+
+      if (column_position == numbers::invalid_unsigned_int)
+        advection_solver_controls.push_back(std::make_pair(column_name,std::vector<SolverControl>(1,solver_control)));
+      else
+        advection_solver_controls[column_position].second.push_back(solver_control);
+    }
+
+    template <int dim>
+    std::pair<std::string,std::string>
+    GlobalStatistics<dim>::execute (TableHandler &statistics)
+    {
+      unsigned int nonlinear_iterations = solver_controls_cheap.size();
+
+      for (unsigned int column=0; column<advection_solver_controls.size(); ++column)
+        nonlinear_iterations = std::max(nonlinear_iterations, static_cast<unsigned int> (advection_solver_controls[column].second.size()));
```

Comment:
I don't think I understand how these four lines generate the number of nonlinear iterations. Can you add a comment?